### PR TITLE
Introduce CCR stats endpoint

### PR DESCRIPTION
**Pull request by jasontedor**
This commit introduces the CCR stats endpoint which provides shard-level stats on the status of CCR follower tasks.

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            }
+            builder.startObject();
+            {
+                for (final Map.Entry<String, Map<Integer, TransportCcrStatsAction.TaskResponse>> index : taskResponsesByIndex.entrySet()) {
+                    builder.startObject(index.getKey());
+                    {
+                        for (final Map.Entry<Integer, TransportCcrStatsAction.TaskResponse> shard : index.getValue().entrySet()) {
+                            builder.startObject(Integer.toString(shard.getKey()));
+                            {
+                                final ShardFollowNodeTask.Status status = shard.getValue().status();
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: why can't we use the status toXContent?

Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: Because there are slight differences in what we present on the endpoint and the status x-content (which is there to enable round-tripping status to/from x-content). For example: the endpoint doesn't present the index metadata version, and presents successful/failed fetches and bulk operations and total/successful.

**Comment by jasontedor** [comment]
Your assumptions are correct.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java (hunk 1 above)
Comment (bleskes) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: re index metadata version - why not present everything we have? the purpose of status object is to report information about the running task, which I think is identical with the purpose of this api?

> presents successful/failed fetches and bulk operations as total/successful.

We can change the status class to track these if you feel it's more natural. It's not a part of the task run time behavior so nothing stops us to align these.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            builder.endObject();
+            return builder;
+        }
+    }
+
+    public static class TasksRequest extends BaseTasksRequest<TasksRequest> {
+
+        private String indexName;
+
+        public void setIndexName(final String indexName) {
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: I think we should follow the same patterns as an indices stats request - i.e., we should except a list of indices or a wild card pattern.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java [hunk 3]
```diff
@@ -59,29 +66,56 @@
...
     private final TimeValue idleShardChangesRequestDelay;
     private final BiConsumer<TimeValue, Runnable> scheduler;
+    private final LongSupplier relativeTimeProvider;
 
-    private volatile long lastRequestedSeqno;
-    private volatile long leaderGlobalCheckpoint;
+    private final ReleasableLock readLock;
+    private final ReleasableLock writeLock;
+
+    {
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: I don't like the locking semantics - read/ write for stats updates and fully synced for class execution. I think we should stick with one. I understand you did it to make sure stats reading is consistent. My vote goes to the simplicity of synchronized access and making the status creation method synchronized as well. It's super light.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportCcrStatsAction.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        return new TaskResponse(in);
+    }
+
+    @Override
+    protected void taskOperation(
+            final CcrStatsAction.TasksRequest request, final ShardFollowNodeTask task, final ActionListener<TaskResponse> listener) {
+        listener.onResponse(new TaskResponse(task.getFollowShardId(), task.getStatus()));
+    }
+
+    public static class TaskResponse implements Writeable {
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportCcrStatsAction.java]: why is this not together with the other request/response object for this action? I'd rather have them together (here or in the action)

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/ShardChangesIT.java [hunk 5]
```diff
@@ -490,7 +491,7 @@ public void testFollowIndex_lowMaxTranslogBytes() throws Exception {
                 ShardFollowNodeTask.Status status = (ShardFollowNodeTask.Status) taskInfo.getStatus();
                 assertThat(status, notNullValue());
                 assertThat("incorrect global checkpoint " + shardFollowTaskParams,
-                    status.getFollowerGlobalCheckpoint(),
+                    status.followerGlobalCheckpoint(),
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/ShardChangesIT.java]: just wondering (I don't care much) - why did you move from the get* method naming?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks @jasontedor . I left some comments. I also miss a test for the TransportCcrStatsAction (the only thing I see that does that's not covered elsewhere is the matching of tasks) and a rest layer test.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportCcrStatsAction.java (hunk 4 above)
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportCcrStatsAction.java]: I pushed 42c309dec588c8167a61742f8f644a88cb851cb1.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java (hunk 1 above)
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: > re index metadata version - why not present everything we have? 

This does not seem useful from an end-user perspective but we can add in the interest of simplification.

> We can change the status class to track these if you feel it's more natural. It's not a part of the task run time behavior so nothing stops us to align these.

Tracking successful/failed feels more natural internally, presenting total/successful feels more natural but we can present successful/failed in the interest of simplification.

I pushed 55300a8fb20dcb74e7333652869751438dc0cf31.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java (hunk 2 above)
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: I think this will require some effort. I will look into this tomorrow.

Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: @bleskes I pushed 2ee14e31d8230afad6b34f224c03ed97a0cf66cd. I am not a fan of it since the `TasksRequest#match` method is now a lie so that we can defer index resolution to the transport action. This is a limitation of the tasks API that puts us here. I do not think we should block CCR on addressing this so we have this workaround for now. I dislike it but maybe it's tolerable for now to give us an API consistent with others. Let me know what you think.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        public boolean match(final Task task) {
+            /*
+             * This is a limitation of the current tasks API. When the transport action is executed, the tasks API invokes this match method
+             * to find the tasks on which to execute the task-level operation (see TransportTasksAction#nodeOperation and
+             * TransportTasksAction#taskOperation). If we do the matching here, then we can not match index patterns. Therefore, we defer
+             * deciding whether or not the task matches the request to the transport action (see TransportCcrStatsAction#taskOperation)
+             * where we can decide on the basis of the cluster state whether or not the task matches any of the index patterns in the
+             * request.
+             */
+            return task instanceof ShardFollowNodeTask;
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: I believe this can be an assertion now? we filter for the task instance type in the transport action.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java [hunk 7]
```diff
@@ -59,29 +66,47 @@
     private final TimeValue retryTimeout;
     private final TimeValue idleShardChangesRequestDelay;
     private final BiConsumer<TimeValue, Runnable> scheduler;
+    private final LongSupplier relativeTimeProvider;
 
-    private volatile long lastRequestedSeqno;
     private volatile long leaderGlobalCheckpoint;
-
+    private volatile long leaderMaxSeqNo;
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: I think these don't need to be volatile any more, now that we read under lock.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java [hunk 8]
```diff
@@ -185,9 +210,30 @@ private void sendShardChangesRequest(long from, int maxOperationCount, long maxR
     }
 
     private void sendShardChangesRequest(long from, int maxOperationCount, long maxRequiredSeqNo, AtomicInteger retryCounter) {
+        final long startTime = relativeTimeProvider.getAsLong();
         innerSendShardChangesRequest(from, maxOperationCount,
-            response -> handleReadResponse(from, maxRequiredSeqNo, response),
-            e -> handleFailure(e, retryCounter, () -> sendShardChangesRequest(from, maxOperationCount, maxRequiredSeqNo, retryCounter)));
+                response -> {
+                    synchronized (ShardFollowNodeTask.this) {
+                        // noinspection NonAtomicOperationOnVolatileField
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: see comment about these not needing to be volatile, meaning we can get rid of this suppression.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestCcrStatsAction.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+import org.elasticsearch.rest.action.RestToXContentListener;
+import org.elasticsearch.xpack.ccr.action.CcrStatsAction;
+
+import java.io.IOException;
+
+public class RestCcrStatsAction extends BaseRestHandler {
+
+    public RestCcrStatsAction(final Settings settings, final RestController controller) {
+        super(settings);
+        controller.registerHandler(RestRequest.Method.GET, "/_xpack/ccr/_stats", this);
```
Comment (bleskes) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestCcrStatsAction.java]: heads up - this will change (or can be changed now)

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks for the extra iterations @jasontedor . Code so far LGTM. I would still love to see a rest test but I presume this is just still in the works.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java (hunk 6 above)
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: I would support throwing `UnsupportedOperationException`.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java (hunk 7 above)
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: You're right. With a little extra synchronization, it is not needed on any of the variables here. I pushed 45aba557b704e3579469e4d4ffa69fe11b3b4586.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java (hunk 6 above)
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrStatsAction.java]: I pushed b9c5429fe8330b32add40c0c33cd4c764eecc018.

**Comment by jasontedor** [comment]
@bleskes Would you take another look?

**Review by bleskes — APPROVED** [review]
